File: 2023/scala/src/main/scala/day04.scala
package day04

import locations.Directory.currentDir
import inputs.Input.loadFileSync
import scala.compiletime.ops.int

@main def part1: Unit =
  println(s"The solution is ${part1(loadInput())}")

@main def part2: Unit =
  println(s"The solution is ${part2(loadInput())}")

def loadInput(): String = loadFileSync(s"$currentDir/../../../../input/day04")

case class ScratchCard(
    cardNumber: Int,
    winning: List[Int],
    mine: List[Int]
):
  def winningCount: Int =
    winning.filter(num => mine.contains(num)).size

  def updateCopies(copies: Map[Int, Int]): Map[Int, Int] =
    (cardNumber + 1 to cardNumber + winningCount).foldLeft(copies)(
      (copies, card) => {

        copies.get(cardNumber) match {
          case Some(times) =>
            copies.updatedWith(card) {
              case Some(v) => Some(v + times)
              case None    => Some(1)
            }
          case None =>
            copies
        }
      }
    )

def getCards(input: String) =
  (for line <- input.linesIterator yield line match {
    case s"Card $cardNumber: $winning | $mine" => {
      ScratchCard(
        cardNumber.trim().toInt,
        winning.split(" ").filter(_.nonEmpty).map(_.trim().toInt).toList,
        mine.split(" ").filter(_.nonEmpty).map(_.trim().toInt).toList
      )
    }
  }).toList

def pow(num: Int, exponent: Int) =
  (1 until exponent).foldLeft(1) { (t, _) => t * 2 }

def part1(input: String): String =

  val cards = getCards(input)

  cards
    .foldLeft(0) { (points, sc) =>
      points +
        (if sc.winningCount == 0 then 0
         else pow(2, sc.winningCount))
    }
    .toString()

def part2(input: String): String =

  val cards = getCards(input)
  val initialCopy = (1 to cards.size).map(_ -> 1).toMap

  val copies = (1 to cards.size).foldLeft(initialCopy) {(copies, k) =>
    copies.get(k) match {
      case Some(_) => (cards(k - 1).updateCopies(copies))
      case None    => copies
    }
  }

  copies.values.sum.toString()

@main def part12: Unit =

  val input = loadInput()
  val cards = getCards(input)

  val totalPoints = cards
    .foldLeft(0) { (points, sc) =>
      points +
        (if sc.winningCount == 0 then 0
         else pow(2, sc.winningCount))
    }
  
  val initialCopy = (1 to cards.size).map(_ -> 1).toMap

  val copies = (1 to cards.size).foldLeft(initialCopy) {(copies, k) =>
    copies.get(k) match {
      case Some(_) => (cards(k - 1).updateCopies(copies))
      case None    => copies
    }
  }

  println(s"Solution for part1 is $totalPoints")
  println(s"Solution for part2 is ${copies.values.sum}")

File: 2023/haskell/src/Day04.hs
module Day04 (part1, part2) where

import qualified Data.Map as Map
import qualified Data.Maybe as Maybe

data Card = Card {number :: Int, winning :: [Int], mine :: [Int]} deriving (Show)

createCard :: String -> Card
createCard line =
  Card {number = num, winning = winningNums, mine = mineNums}
  where
    (_ : digits : xs) = words line
    num = read (init digits) :: Int
    (ws, "|" : ms) = span (/= "|") xs
    winningNums = map (read :: String -> Int) ws
    mineNums = map (read :: String -> Int) ms

matchingNumbers :: Card -> Int
matchingNumbers Card {winning = ws, mine = ms} =
  length $ filter (`elem` ms) ws

winningCount :: Card -> Int
winningCount card =
  let count = matchingNumbers card
   in if count == 0 then 0 else iterate (* 2) 1 !! (count - 1)

updateCopies :: [Card] -> Map.Map Int Int -> Int -> Map.Map Int Int
updateCopies cards m i =
  foldl
    ( \m1 n ->
        let copies = Maybe.fromMaybe 0 (Map.lookup n m)
         in Map.insert n (copies + times) m1
    )
    m
    copiesToAdd
  where
    card@Card {number = cardNo} = cards !! (i - 1)
    count = matchingNumbers card
    times = Maybe.fromMaybe 0 (Map.lookup i m)
    copiesToAdd = map (+ cardNo) [1 .. count]

part1 :: String -> IO ()
part1 input = do
  print $ sum $ map (winningCount . createCard) $ lines input

part2 :: String -> IO ()
part2 input = do
  print $ sum $ Map.map id $ foldl (updateCopies cards) initialCopies indices
  where
    cards = map createCard (lines input)
    indices = [1 .. length cards]
    initialCopies = Map.fromList [(index, 1) | index <- indices]